[Merged by Bors] - API phase 11: cleanup old API and protoc code

lrettig opened the pull request:
## Motivation
See #1764 
Closes #2077, Closes #2078, Closes #2153, Closes #411, Closes #2163, Closes #2145, Closes #2168

## Changes
- Removes old GRPC API and HTTP-JSON gateway code
- Removes protobuf code
- Updates tests to use new API
- Lots of cleanup of the system tests

## Test Plan
No new tests required, this does not add any new functionality. Affected tests are updated/removed.

## TODO
- [x] Remove old API code
- [x] Update tests to use new API instead
- [x] Remove all protobuf-related code and references
- [x] Update README
- [x] Update to API v1.0.0, adding GatewayService to new API for poet (#2133)
- [x] Implement GatewayService (#2156)
- [x] Update poet (https://github.com/spacemeshos/poet/issues/98)
- [x] Fix p2p tests broadcast issue
- [x] Update hardcoded go-spacemesh docker images in configs
- [x] Open issue to replace all `curl`/json http gateway server API interactions in tests with python grpc library (#2185)
- [x] Update system tests (including tx_generator WalletAPI, test_p2p) to use new API
- [x] Make sure spacecraft has been updated to factor in new ports and commandline flags (https://github.com/spacemeshos/spacecraft/issues/1)
- [x] Update hardcoded poet docker image spacemeshos/poet:develop, fix in configs (needs https://github.com/spacemeshos/poet/pull/105)
- [x] Merge #2157 
- [x] Merge #2159
- [x] Merge #2169

lrettig committed:
Remove old API code

Removes old GRPC and HTTP-JSON gateway code and related tests
Removes old API from multinode and p2p Cmds
Renames API-related config/cmdline flags to remove "new" for the new API

lrettig committed:
Remove protobuf

and its related scripts and configs.
Finally! It's no longer needed since now we import the code as a go
library directly.
Update configs to use new API, and to not try to run the old API.

lrettig committed:
Rewrite app_test to use new API

Re-add new API to multi_node
Update harness port to not conflict with new API port
Allow app to store references to individual API services so that they
can be used directly in tests

lrettig committed:
Remove some unused interfaces

lrettig committed:
go mod tidy

lrettig committed:
Remove lingering genproto reference

lrettig committed:
Remove data elements from GlobalStateService

They're unused

lrettig committed:
go mod tidy

lrettig commented:
bors try

lrettig committed:
Fix docker build

lrettig committed:
Merge branch 'develop' into api-cleanup

lrettig committed:
Upgrade testify module

Upgrade to github.com/stretchr/testify@v1.6.1

lrettig committed:
Cleanup poet harness tests

Add lsof package to avoid error reported in
https://github.com/spacemeshos/poet/issues/102 on Alpine linux

lrettig committed:
Fix small merge issues

Caused by bringing in latest code from develop

lrettig committed:
Fix tests

Update command line flags. This should stop tests from running forever,
but the underlying issue is unaddressed, see
https://github.com/spacemeshos/go-spacemesh/issues/2151.

lrettig committed:
Fix error in TestSpacemeshApp_JsonFlags

lrettig committed:
Remove unused arg from testArgs function

lrettig committed:
Cleanup tests further

Always shutdown correctly
Fix typos

lrettig committed:
Properly set up and clean up tempdirs

lrettig committed:
Update grpc and json ports

lrettig committed:
More minor test cleanup

sync.once is probably not necessary here

lrettig committed:
Misc minor test cleanup

lrettig committed:
Add a timeout while waiting for nonce

Prevent the infinite loop we're currently stuck in

lrettig committed:
Capture poet harness stdout/stderr

lrettig committed:
Better handle poet logs

Combine stderr and stdout into one reader, catch errors early and
report them on an error channel

lrettig committed:
Improve timeout and error handling

Add top level timeout for setup, and fail fast on poet harness errors
too

lrettig committed:
Update modules

Pull in the latest API release.
Some existing google modules are incompatible with grpc >= v1.30.0, so
update the affected google modules as well.

lrettig committed:
Return current and projected account state

from GlobalStateService

lrettig committed:
Fix reward layer data type

lrettig committed:
Return current and projected account state

everywhere, in GlobalStateService
Factor out the code that generates this so as to DRY

lrettig committed:
Remove AppAddress

lrettig committed:
Fix layer number types in MeshService

lrettig committed:
Update handling of error levels to match API

lrettig committed:
Silence warnings about unused args

lrettig committed:
Add unimplemented endpoint to SmesherService

lrettig committed:
Silence warnings about unused args

lrettig committed:
Silence warning about unused args in two more svcs

lrettig committed:
Resolve all warnings in TransactionService

shadowed variable names and unused args

lrettig committed:
Begin cleaning up API tests

Fix layer numbers, receipts

lrettig committed:
Create a Mempool API for use in the API code

Wrap mempool related methods in an interface like we do for other types
of interaction with the node inside the API. Makes testing easier.

lrettig committed:
Finish updating grpcserver tests

Use the new Mempool API to create a mock mempool object and use it to
test both current and projected state

lrettig committed:
go fmt

lrettig committed:
Add gateway service for poet

lrettig committed:
Update node test to match latest API

lrettig committed:
Upgrade to latest API including new service

lrettig committed:
Upgrade API

This build includes fixes for some lingering Layer data type issues

lrettig committed:
Fixes related to layer type

lrettig committed:
More test fixes for layer issues

lrettig committed:
Merge branch 'develop' into api-v1.0.0

lrettig committed:
Fix another test

lrettig committed:
Fix TransactionsState_SubmitThenStream test

New mempool mocker needs to report new tx added to mempool

lrettig committed:
Few more minor test fixes

lrettig committed:
Make linter happy

lrettig committed:
Remove comment

lrettig committed:
Merge branch 'api-v1.0.0' into gateway-service

lrettig committed:
Update to latest API branch

lrettig committed:
Finish implementing GatewayService

Add a basic test

lrettig committed:
Merge branch 'develop' into api-cleanup

lrettig committed:
go mod tidy

lrettig committed:
Merge branch 'gateway-service' into api-cleanup

lrettig committed:
Remove unused PeerCounter from GlobalStateService

lrettig committed:
Fix a test, run go mod tidy

Update another test to match latest API and check current (rather than
projected) state

lrettig committed:
Cleanup protobuf modules

lrettig committed:
Add config for GatewayService

Note that the API does not currently define a JSON gateway service for
this service, so we don't attempt to enable it yet

Also alphabetizes the list of services to make managing it easier

lrettig committed:
Merge branch 'gateway-service' into api-cleanup

lrettig committed:
Activate GatewayService for test

lrettig committed:
Bugfix

lrettig committed:
Merge branch 'gateway-service' into api-cleanup

lrettig committed:
Add GatewayService json gateway

lrettig committed:
Merge branch 'gateway-service' into api-cleanup

lrettig committed:
Update endpoints for system tests

lrettig committed:
Misc test cleanup

Finish setup much more quickly by checking projected state
Simplify test-checking loop, improve log output

lrettig committed:
Shutdown before running more tests

And wrap shutdown routine so we only run it once

lrettig committed:
Unset jsonlog in test

lrettig commented:
bors try

lrettig committed:
Update grpc command line flags in k8s config

lrettig commented:
bors try

lrettig committed:
More test cleanup and improvements

Improve the way we handle shutdown in app_test
Use temp dirs for state for tests that actually run nodes in node_tests

lrettig commented:
bors try

lrettig committed:
Merge branch 'develop' into api-cleanup

lrettig committed:
Add GRPC API services back to p2p runner

Misc test cleanup
Fix a couple of typos

Messages broadcast in p2p test are still hitting the wrong endpoint
(BroadcastPoet) as the new API doesn't have a simple broadcast message
endpoint, fixing this is TODO

lrettig committed:
Run with updated poet image

lrettig commented:
bors try

lrettig committed:
Remove genproto from CI

lrettig committed:
Fix P2P tests

- use PoetProof protocol instead of bespoke api test protocol
- base64 encode messages as API is expecting binary data
- add back api_test_helper to override this protocol and mark all
  gossip messages as approved
- change hardcoded docker image names (temporarily) for purposes of
  testing outside bors, see #2163 for more info on why this is necessary

lrettig committed:
Update one more hardcoded image name

P2P test is now passing locally, yay!
Fix a typo

lrettig commented:
bors try

lrettig committed:
Fix docker image names

Harmonize naming of 'clientv2' vs 'client_v2' to resolve #2163

lrettig committed:
Misc test cleanup

Modify wallet_api to work correctly with new API via json gateway,
sending tx data as base64 encoded byte array, etc.
Sleep long enough during broadcast_unknown_protocol test to allow
bootstrap to occur
Fix misc typos

lrettig commented:
bors try

lrettig commented:
bors try

lrettig commented:
bors try

lrettig committed:
Fix faulty logic in version checker

Clean up an error message

lrettig committed:
Fix P2P bootstrap version comparison logic

Adds a regression test, copied from #2169
Use a different docker image for clientv2 test, one that has an _older_
(not newer) version (0.0.1 vs 0.0.2) for P2P bootstrap version test, and
that accepts the same command line args
And don't override this image name in test setup

lrettig committed:
Bump P2P version to 0.0.2

This should allow test_diff_client_ver test to reject a node with v0.0.1

lrettig committed:
Logging cleanup

lrettig committed:
Update clientv2 image

to match that used in p2p test

lrettig committed:
Remove clientv2 test, cleanup logs

See #2168 for more info on clientv2 test issue

lrettig committed:
Update API to work with TransactionService

lrettig committed:
Merge branch 'develop' into api-cleanup

lrettig committed:
Fix transaction API

systemtest-hare-mining is now passing

lrettig committed:
Clean up logging and loop logic

lrettig committed:
Revert "Remove clientv2 test, cleanup logs"

And "Update clientv2 image"

This reverts commits e3f1840acc2c8d482deaa8eceac27dac4658bb1c and 864b1e833301c44b896aa02b1878f7b734b3dd18.

Taking a different approach here where we allow and replace clientv2
images

lrettig committed:
Add noreplace property to p2p image config

lrettig committed:
Cleanup log output and whitespace

lrettig committed:
Remove an infinite loop

lrettig committed:
Bugfix in logic

lrettig commented:
This is basically done, and all system tests are now passing for me locally. It's blocking on the following things in roughly the following order:

- #2157 needs to be reviewed and merged (@y0sher, @antonlerner 🙏)
- #2159 needs to be reviewed and merged (@y0sher, @antonlerner 🙏)
- the PRs I opened against poet need to be reviewed and merged, most crucially, https://github.com/spacemeshos/poet/pull/105 (@moshababo 🙏)
- once that's done, need to fix hardcoded poet docker image here
- need to discuss #2168 and resolve that

Then this is good to go.

lrettig committed:
Pin to older version of poet

Before upgrade to new API

Cherry-picks 387e57ddd01d78f99114d7fbf0718d3fc6b72d68 to switch to
pinned version of poet

lrettig committed:
Merge branch 'develop' into api-cleanup

lrettig committed:
Clarify what v2 means in version test

lrettig committed:
Merge branch 'develop' into api-cleanup

lrettig committed:
tidy modules

lrettig committed:
More post-merge cleanup

lrettig committed:
More module cleanup

lrettig committed:
Merge branch 'develop' into api-cleanup

lrettig commented:
bors try

lrettig commented:
bors try-

lrettig commented:
bors try

antonlerner commented on README.md:
````diff
@@ -167,12 +167,12 @@ Assuming tn1.json is a testnet config file saved in the same directory as go-spa
 
 ```bash
-./go-spacemesh --grpc-server --json-server --tcp-port [a_port] --config [configFileLocation] -d [nodeDataFilesPath] --coinbase [coinbase_account] --start-mining --post-datadir [dir_for_post_data]
+./go-spacemesh --tcp-port [a_port] --config [configFileLocation] -d [nodeDataFilesPath] --coinbase [coinbase_account] --start-mining --post-datadir [dir_for_post_data]
````
we don't need GRPC servers anymore?

lrettig replied:
It's necessary if you're going to connect smapp or the CLIWallet to it, but it's not necessary if you just want to run a node on the command line, which is what these instructions are for. We still need to separately document the API (see #2083) since it's a bit confusing how to use it properly.

antonlerner commented on cmd/multi_node_sim/config.toml:
```diff
@@ -46,8 +46,8 @@ bootnodes = [] # example : spacemesh://j7qWfWaJRVp25ZsnCu9rJ4PmhigZBtesB4YmQHqqP
 json-server = true
-grpc-port = 9091
-json-port = 9090
+grpc-port = 9092
```
why did the port change?

antonlerner commented on config.toml:
```diff
@@ -45,10 +45,10 @@ bootnodes = [] # example : spacemesh://j7qWfWaJRVp25ZsnCu9rJ4PmhigZBtesB4YmQHqqP
 json-server = true
-grpc-port = 9091
-json-port = 9090
+grpc-port = 9092
```
why change the port?

lrettig replied:
We don't need to change the port, but the API services are totally different, so it's sort of a way of requiring any clients (tests, etc.) that were talking to the old API service to explicitly choose to talk to the new one, does that make sense? Let me know what you think.

antonlerner commented:
could you also remove the print that says `not reporting status update as no one is listening` please?

antonlerner commented on cmd/node/multi_node.go:
```diff
@@ -251,7 +259,9 @@ func StartMultiNode(numOfinstances, layerAvgSize int, runTillLayer uint32, dbPat
 	}
 	pubsubAddr := "tcp://localhost:55666"
-	events.InitializeEventReporter(pubsubAddr)
+	if err := events.InitializeEventReporter(pubsubAddr); err != nil {
```
make sure you close this somewhere when the node is done

lrettig replied:
good catch, fixed

lrettig commented:
> could you also remove the print that says `not reporting status update as no one is listening` please?

I'll change this from INFO to DEBUG, does that make sense? It's sort of useful as a debug message so that you know why not everything is being reported, also because of #2085

lrettig committed:
Merge branch 'develop' into api-cleanup

lrettig committed:
Close event reporter in multi node

Spotter: Anton

lrettig committed:
Improve reporter logger

Downgrade verbose log messages to debug
Log activation (as other data types are logged)

lrettig commented:
bors try

lrettig committed:
Update pinned poet version

lrettig commented:
bors try-

lrettig commented:
bors try

lrettig committed:
Improve test error handling

lrettig commented:
bors merge

lrettig commented:
bors merge

lrettig commented:
bors merge
